certora/specs/authorizations.spec

```diff
@@ -0,0 +1,263 @@
+using Utils as utils
+
+methods {
+    totalSupply() returns (uint256) envfree
+    balanceOf(address) returns (uint256) envfree
+    allowance(address, address) returns (uint256) envfree
+    owner() returns (address) envfree
+    getUserRoles(address) returns (bytes32) envfree
+    isCapabilityPublic(bytes4) returns (bool) envfree
+    getRolesWithCapability(bytes4) returns (bytes32) envfree
+    doesUserHaveRole(address, uint8) returns (bool) envfree
+    doesRoleHaveCapability(uint8, bytes4) returns (bool) envfree
+    utils.to_bytes4(uint32) returns (bytes4) envfree
+}
+
+
+// AUTHORIZATION STORAGE 
+
+// The authorization variables are: owner, isCapabilityPublic, getUserRoles and getRolesCapability.
+
+// owner variable
+
+rule ownerChanging() {
+    env e;
+    method f; calldataarg args; bytes4 newOwner;
+    address ownerBefore = owner();
+
+    f(e, args);
+
+    address ownerAfter = owner();
+    assert (ownerAfter != ownerBefore =>
+            f.selector == setOwner(address).selector);
+}
+
+rule setOwnerShouldChangeOwner(address newOwner) {
+    env e;
+
+    setOwner(e, newOwner);
+
+    address ownerAfter = owner();
+    assert (ownerAfter == newOwner);
+}
+
+// isCapabilityPublic mapping
+
+rule isCapabilityPublicChanging() {
+    env e;
+    method f; calldataarg args; bytes4 capability;
+    bool capabilityPublicBefore = isCapabilityPublic(capability);
+
+    f(e, args);
+
+    bool capabilityPublicAfter = isCapabilityPublic(capability);
+    assert (capabilityPublicAfter != capabilityPublicBefore =>
+            f.selector == setPublicCapability(bytes4, bool).selector);
```

Comment:
we could also have a rule that check if the arg I the right one

Reply:
Yes, but we can't do that in the same rule because there is no way to decompose an arbitrary `args`. That is why it is done in the next rule: `setPublicCapabilityShouldChangeIsPublicCapability`